crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1871,6 +1935,29 @@ impl<'db> ClassType<'db> {
     }
 }
 
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct InstanceType<'db> {
+    class: ClassType<'db>,
+    known: Option<KnownInstance>,
+}
+
+impl<'db> InstanceType<'db> {
+    pub fn anonymous(class: ClassType<'db>) -> Self {
+        Self { class, known: None }
+    }
+
+    pub fn known(class: ClassType<'db>, known: KnownInstance) -> Self {
+        Self {
+            class,
+            known: Some(known),
+        }
+    }
```

Comment:
I think the `known` method here needs to have a different name, or it conflicts with the name of the field on the class. Rust _lets_ you do that, but I find it pretty confusing to have a method the same name as a field that does quite a differen thing 😆

How about `new_known()` and `new_unknown` for the two constructors? That would also have the nice property of making it obvious that they are parallel to each other